### PR TITLE
make tests/models/sam_hq/test_modeling_sam_hq.py::SamHQModelIntegrati…

### PR DESCRIPTION
**Pull request by sywangyi**
…onTest pass in cuda and xpu

@Rocketknight1

### CODE REVIEW

**Comment by sywangyi** [comment]
fix https://github.com/huggingface/transformers/issues/42890

**Comment by Rocketknight1** [comment]
run-slow: sam_hq

**Comment by Rocketknight1** [comment]
cc @ydshieh! I confirmed that the SamHQ slow tests were failing locally for me before this

**Comment by yao-matrix** [comment]
@ydshieh , could you pls help review this, thx very much.

**Comment by yao-matrix** [comment]
@ydshieh , happy new year! could you help take a look at this PR? Thx very much.

**Comment by ydshieh** [comment]
Hi. Thank you for diving into this.

The checkpoint

https://huggingface.co/syscv-community/sam-hq-vit-base/blob/main/model.safetensors

has the weight `shared_image_embedding.positional_embedding`, so it's not very clear why we need to set seed, as the model is loaded from that checkpoint

> model = SamHQModel.from_pretrained("syscv-community/sam-hq-vit-base")

Do you have any idea what actually causes the issue? Maybe we don't need seed, and the original failure is just the value was wrong  but it's not really changing on each run? cc @Rocketknight1 as you mentioned this particular

https://github.com/huggingface/transformers/issues/42890#issuecomment-3660799382

Anyway, it's too much time consuming to change again, so I am happy to approve and merge as it, but just add a comment to the seed to say it's not really necessary but just for convenience (if it turns out not necessary).

If possible, let's avoid use anything like

> self.assertTrue(torch.allclose(scores[0][0][-1], torch.tensor(0.4482), atol=2e-4))

but `torch.testing.assert_close` instead. But if you have no bandwidth, it's fine.

**Comment by sywangyi** [comment]
see https://github.com/huggingface/transformers/issues/42890 description. 
see https://github.com/sywangyi/transformers/blob/main/src/transformers/models/sam_hq/modeling_sam_hq.py#L1077
randn() is used in positional_embedding, so you need to set_seed() before modeling loading to guarantee the stable output. or else accuracy check will fail.

**Comment by ydshieh** [comment]
@sywangyi I understand randn() is used in positional_embedding. But in the tests, we are not creating a model from scratch but loading from a mode checkpoint 

> model = SamHQModel.from_pretrained("syscv-community/sam-hq-vit-base")

which has the weight 

> shared_image_embedding.positional_embedding

so eventually, the loaded model has the same weight across run.

Unless there is any other random operation involved,  we shouldn't have the results being different across runs.

So my question is:

- Do you observe the results being different across different multiple runs.
  - if this is the case, the root cause might not be the   randn() is used in positional_embedding 
- Or the results are consistent, but just not matching the expected values on the main branch.

**Comment by sywangyi** [comment]
it's "prompt_encoder.shared_embedding.positional_embedding" [https://github.com/huggingface/transformers/blob/main/src/transformers/models/sam_hq/modeling_sam_hq.py#L1238]
this param is random, however, I found set_seed also could not guarantee it's fixed. so maybe we need to fix the param.

change seed value now could make the result being different

> @sywangyi I understand randn() is used in positional_embedding. But in the tests, we are not creating a model from scratch but loading from a mode checkpoint
> 
> > model = SamHQModel.from_pretrained("syscv-community/sam-hq-vit-base")
> 
> which has the weight
> 
> > shared_image_embedding.positional_embedding
> 
> so eventually, the loaded model has the same weight across run.
> 
> Unless there is any other random operation involved, we shouldn't have the results being different across runs.
> 
> So my question is:
> 
> * Do you observe the results being different across different multiple runs.
>   
>   * if this is the case, the root cause might not be the   randn() is used in positional_embedding
> * Or the results are consistent, but just not matching the expected values on the main branch.

**Comment by ydshieh** [comment]
Hi @sywangyi 

> it's "prompt_encoder.shared_embedding.positional_embedding"

Thank you a lot for identifying this ! Great detective work 🙏 .

Let me ping a team member to see what's going on with this missing weight on the hub repository, or it's designed to be so.
I will come back to you as soon as possible.

**Comment by ydshieh** [comment]
I verified and this is more involved which requires a fix in the modeling code (not simply the test file).

We will try to work on this next week 🙏

**Comment by sywangyi** [comment]
PR in modeling has been merged
